// @n8n/ai-workflow-builder/dist/utils/context-builders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildWorkflowSummary = buildWorkflowSummary;
exports.buildWorkflowJsonBlock = buildWorkflowJsonBlock;
exports.buildDiscoveryContextBlock = buildDiscoveryContextBlock;
exports.buildExecutionContextBlock = buildExecutionContextBlock;
exports.buildExecutionSchemaBlock = buildExecutionSchemaBlock;
exports.createContextMessage = createContextMessage;
const messages_1 = require("@langchain/core/messages");
const trim_workflow_context_1 = require("./trim-workflow-context");
function buildWorkflowSummary(workflow) {
    if (workflow.nodes.length === 0) {
        return 'No nodes in workflow';
    }
    const nodeList = workflow.nodes.map((n) => `- ${n.name} (${n.type})`).join('\n');
    return `${workflow.nodes.length} existing nodes:\n${nodeList}`;
}
function buildWorkflowJsonBlock(workflow) {
    const trimmed = (0, trim_workflow_context_1.trimWorkflowJSON)(workflow);
    return [
        '<current_workflow_json>',
        JSON.stringify(trimmed, null, 2),
        '</current_workflow_json>',
        '<note>Large property values may be trimmed. Use get_node_parameter tool for full details.</note>',
    ].join('\n');
}
function buildDiscoveryContextBlock(discoveryContext, includeBestPractices = true) {
    if (!discoveryContext)
        return '';
    const parts = [];
    if (discoveryContext.nodesFound.length > 0) {
        parts.push('Discovered Nodes:');
        discoveryContext.nodesFound.forEach(({ nodeName, version, reasoning, connectionChangingParameters, availableResources }) => {
            const params = connectionChangingParameters.length > 0
                ? ` [Connection params: ${connectionChangingParameters.map((p) => p.name).join(', ')}]`
                : '';
            let resourceInfo = '';
            if (availableResources && availableResources.length > 0) {
                const filteredResources = availableResources
                    .filter((r) => r.value !== '__CUSTOM_API_CALL__')
                    .map((r) => {
                    const ops = r.operations
                        .filter((o) => o.value !== '__CUSTOM_API_CALL__')
                        .map((o) => `"${o.value}"`)
                        .join(', ');
                    return `    - resource="${r.value}" supports operations: [${ops}]`;
                });
                if (filteredResources.length > 0) {
                    resourceInfo = `\n  Available resource/operation combinations:\n${filteredResources.join('\n')}`;
                }
            }
            parts.push(`- ${nodeName} v${version}: ${reasoning}${params}${resourceInfo}`);
        });
    }
    if (includeBestPractices && discoveryContext.bestPractices) {
        parts.push('', 'Best Practices:', discoveryContext.bestPractices);
    }
    return parts.join('\n');
}
function buildExecutionContextBlock(workflowContext) {
    const executionData = workflowContext?.executionData ?? {};
    const executionSchema = workflowContext?.executionSchema ?? [];
    return [
        '<execution_data>',
        JSON.stringify(executionData, null, 2),
        '</execution_data>',
        '',
        '<execution_schema>',
        JSON.stringify(executionSchema, null, 2),
        '</execution_schema>',
    ].join('\n');
}
function buildExecutionSchemaBlock(workflowContext) {
    const executionSchema = workflowContext?.executionSchema ?? [];
    if (executionSchema.length === 0)
        return '';
    return [
        '<execution_schema>',
        JSON.stringify(executionSchema, null, 2),
        '</execution_schema>',
    ].join('\n');
}
function createContextMessage(contextParts) {
    return new messages_1.HumanMessage({ content: contextParts.filter(Boolean).join('\n\n') });
}
