// @n8n/ai-workflow-builder/dist/subgraphs/builder.subgraph.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuilderSubgraph = exports.BuilderSubgraphState = void 0;
const prompts_1 = require("@langchain/core/prompts");
const langgraph_1 = require("@langchain/langgraph");
const errors_1 = require("../errors");
const builder_prompt_1 = require("../prompts/agents/builder.prompt");
const auto_fix_1 = require("../validation/auto-fix");
const checks_1 = require("../validation/checks");
const subgraph_interface_1 = require("./subgraph-interface");
const add_node_tool_1 = require("../tools/add-node.tool");
const connect_nodes_tool_1 = require("../tools/connect-nodes.tool");
const get_node_examples_tool_1 = require("../tools/get-node-examples.tool");
const remove_connection_tool_1 = require("../tools/remove-connection.tool");
const remove_node_tool_1 = require("../tools/remove-node.tool");
const rename_node_tool_1 = require("../tools/rename-node.tool");
const validate_structure_tool_1 = require("../tools/validate-structure.tool");
const coordination_1 = require("../types/coordination");
const cache_control_1 = require("../utils/cache-control");
const context_builders_1 = require("../utils/context-builders");
const operations_processor_1 = require("../utils/operations-processor");
const state_reducers_1 = require("../utils/state-reducers");
const subgraph_helpers_1 = require("../utils/subgraph-helpers");
exports.BuilderSubgraphState = langgraph_1.Annotation.Root({
    workflowJSON: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => ({ nodes: [], connections: {}, name: '' }),
    }),
    userRequest: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => '',
    }),
    workflowContext: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
    }),
    discoveryContext: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => null,
    }),
    messages: (0, langgraph_1.Annotation)({
        reducer: (x, y) => x.concat(y),
        default: () => [],
    }),
    workflowOperations: (0, langgraph_1.Annotation)({
        reducer: (x, y) => {
            if (y === null)
                return [];
            if (!y || y.length === 0)
                return x ?? [];
            return [...(x ?? []), ...y];
        },
        default: () => [],
    }),
    cachedTemplates: (0, langgraph_1.Annotation)({
        reducer: state_reducers_1.cachedTemplatesReducer,
        default: () => [],
    }),
});
class BuilderSubgraph extends subgraph_interface_1.BaseSubgraph {
    name = 'builder_subgraph';
    description = 'Constructs workflow structure: creating nodes and connections';
    config;
    create(config) {
        this.config = config;
        const includeExamples = config.featureFlags?.templateExamples === true;
        const baseTools = [
            (0, add_node_tool_1.createAddNodeTool)(config.parsedNodeTypes),
            (0, connect_nodes_tool_1.createConnectNodesTool)(config.parsedNodeTypes, config.logger),
            (0, remove_node_tool_1.createRemoveNodeTool)(config.logger),
            (0, remove_connection_tool_1.createRemoveConnectionTool)(config.logger),
            (0, rename_node_tool_1.createRenameNodeTool)(config.logger),
            (0, validate_structure_tool_1.createValidateStructureTool)(config.parsedNodeTypes),
        ];
        const tools = includeExamples
            ? [...baseTools, (0, get_node_examples_tool_1.createGetNodeConnectionExamplesTool)(config.logger)]
            : baseTools;
        const toolMap = new Map(tools.map((bt) => [bt.tool.name, bt.tool]));
        const systemPrompt = prompts_1.ChatPromptTemplate.fromMessages([
            [
                'system',
                [
                    {
                        type: 'text',
                        text: (0, builder_prompt_1.buildBuilderPrompt)(),
                        cache_control: { type: 'ephemeral' },
                    },
                ],
            ],
            ['placeholder', '{messages}'],
        ]);
        if (typeof config.llm.bindTools !== 'function') {
            throw new errors_1.LLMServiceError('LLM does not support tools', {
                llmModel: config.llm._llmType(),
            });
        }
        const agent = systemPrompt.pipe(config.llm.bindTools(tools.map((bt) => bt.tool)));
        const callAgent = async (state) => {
            (0, cache_control_1.applySubgraphCacheMarkers)(state.messages);
            const response = await agent.invoke({
                messages: state.messages,
            });
            return { messages: [response] };
        };
        const shouldContinue = (0, subgraph_helpers_1.createStandardShouldContinue)();
        const subgraph = new langgraph_1.StateGraph(exports.BuilderSubgraphState)
            .addNode('agent', callAgent)
            .addNode('tools', async (state) => await (0, subgraph_helpers_1.executeSubgraphTools)(state, toolMap))
            .addNode('process_operations', operations_processor_1.processOperations)
            .addEdge('__start__', 'agent')
            .addConditionalEdges('agent', shouldContinue)
            .addEdge('tools', 'process_operations')
            .addEdge('process_operations', 'agent');
        return subgraph.compile();
    }
    transformInput(parentState) {
        const userRequest = (0, subgraph_helpers_1.extractUserRequest)(parentState.messages);
        const contextParts = [];
        contextParts.push('=== USER REQUEST ===');
        contextParts.push(userRequest);
        if (parentState.discoveryContext) {
            contextParts.push('=== DISCOVERY CONTEXT ===');
            contextParts.push((0, context_builders_1.buildDiscoveryContextBlock)(parentState.discoveryContext, true));
        }
        contextParts.push('=== CURRENT WORKFLOW ===');
        if (parentState.workflowJSON.nodes.length > 0) {
            contextParts.push((0, context_builders_1.buildWorkflowJsonBlock)(parentState.workflowJSON));
        }
        else {
            contextParts.push('Empty workflow - ready to build');
        }
        const schemaBlock = (0, context_builders_1.buildExecutionSchemaBlock)(parentState.workflowContext);
        if (schemaBlock) {
            contextParts.push('=== AVAILABLE DATA SCHEMA ===');
            contextParts.push(schemaBlock);
        }
        const contextMessage = (0, context_builders_1.createContextMessage)(contextParts);
        return {
            userRequest,
            workflowJSON: parentState.workflowJSON,
            workflowContext: parentState.workflowContext,
            discoveryContext: parentState.discoveryContext,
            messages: [contextMessage],
            cachedTemplates: parentState.cachedTemplates,
        };
    }
    transformOutput(subgraphOutput, _parentState) {
        let workflowJSON = subgraphOutput.workflowJSON;
        let autoFixSummary;
        if (this.config?.parsedNodeTypes) {
            const violations = (0, checks_1.validateConnections)(workflowJSON, this.config.parsedNodeTypes);
            const autoFixResult = (0, auto_fix_1.autoFixConnections)(workflowJSON, this.config.parsedNodeTypes, violations);
            if (autoFixResult.fixed.length > 0) {
                workflowJSON = {
                    ...workflowJSON,
                    connections: autoFixResult.updatedConnections,
                };
                autoFixSummary = `Auto-fixed ${autoFixResult.fixed.length} connection(s): ${autoFixResult.fixed
                    .map((fix) => `${fix.sourceNodeName} → ${fix.targetNodeName}`)
                    .join(', ')}`;
            }
        }
        const nodes = workflowJSON.nodes;
        const connections = workflowJSON.connections;
        const connectionCount = Object.values(connections).flat().length;
        const builderSummary = subgraphOutput.messages
            .slice()
            .reverse()
            .find((m) => m.content &&
            (!('tool_calls' in m) ||
                !m.tool_calls ||
                (m.tool_calls && Array.isArray(m.tool_calls) && m.tool_calls.length === 0)));
        const summaryText = typeof builderSummary?.content === 'string' ? builderSummary.content : undefined;
        const summary = autoFixSummary
            ? `Created ${nodes.length} nodes with ${connectionCount} connections. ${autoFixSummary}`
            : `Created ${nodes.length} nodes with ${connectionCount} connections`;
        const logEntry = {
            phase: 'builder',
            status: 'completed',
            timestamp: Date.now(),
            summary,
            output: summaryText,
            metadata: (0, coordination_1.createBuilderMetadata)({
                nodesCreated: nodes.length,
                connectionsCreated: connectionCount,
                nodeNames: nodes.map((n) => n.name),
            }),
        };
        return {
            workflowJSON,
            workflowOperations: subgraphOutput.workflowOperations ?? [],
            coordinationLog: [logEntry],
            cachedTemplates: subgraphOutput.cachedTemplates,
        };
    }
}
exports.BuilderSubgraph = BuilderSubgraph;
