// @n8n/ai-workflow-builder/dist/subgraphs/configurator.subgraph.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfiguratorSubgraph = exports.ConfiguratorSubgraphState = void 0;
const prompts_1 = require("@langchain/core/prompts");
const langgraph_1 = require("@langchain/langgraph");
const errors_1 = require("../errors");
const configurator_prompt_1 = require("../prompts/agents/configurator.prompt");
const subgraph_interface_1 = require("./subgraph-interface");
const get_node_examples_tool_1 = require("../tools/get-node-examples.tool");
const get_node_parameter_tool_1 = require("../tools/get-node-parameter.tool");
const get_resource_locator_options_tool_1 = require("../tools/get-resource-locator-options.tool");
const update_node_parameters_tool_1 = require("../tools/update-node-parameters.tool");
const validate_configuration_tool_1 = require("../tools/validate-configuration.tool");
const coordination_1 = require("../types/coordination");
const langchain_1 = require("../types/langchain");
const cache_control_1 = require("../utils/cache-control");
const context_builders_1 = require("../utils/context-builders");
const operations_processor_1 = require("../utils/operations-processor");
const rlc_prefetch_1 = require("../utils/rlc-prefetch");
const state_reducers_1 = require("../utils/state-reducers");
const subgraph_helpers_1 = require("../utils/subgraph-helpers");
exports.ConfiguratorSubgraphState = langgraph_1.Annotation.Root({
    workflowJSON: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => ({ nodes: [], connections: {}, name: '' }),
    }),
    workflowContext: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
    }),
    instanceUrl: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => '',
    }),
    userRequest: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => '',
    }),
    discoveryContext: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => null,
    }),
    messages: (0, langgraph_1.Annotation)({
        reducer: (x, y) => x.concat(y),
        default: () => [],
    }),
    workflowOperations: (0, langgraph_1.Annotation)({
        reducer: (x, y) => {
            if (y === null)
                return [];
            if (!y || y.length === 0)
                return x ?? [];
            return [...(x ?? []), ...y];
        },
        default: () => [],
    }),
    cachedTemplates: (0, langgraph_1.Annotation)({
        reducer: state_reducers_1.cachedTemplatesReducer,
        default: () => [],
    }),
    prefetchedRLCOptions: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => [],
    }),
});
class ConfiguratorSubgraph extends subgraph_interface_1.BaseSubgraph {
    name = 'configurator_subgraph';
    description = 'Configures node parameters after structure is built';
    agent;
    toolMap;
    instanceUrl = '';
    nodeTypes = [];
    resourceLocatorCallback;
    logger;
    create(config) {
        this.instanceUrl = config.instanceUrl ?? '';
        this.nodeTypes = config.parsedNodeTypes;
        this.resourceLocatorCallback = config.resourceLocatorCallback;
        this.logger = config.logger;
        const includeExamples = config.featureFlags?.templateExamples === true;
        const parameterUpdaterLLM = config.llmParameterUpdater ?? config.llm;
        const baseTools = [
            (0, update_node_parameters_tool_1.createUpdateNodeParametersTool)(config.parsedNodeTypes, parameterUpdaterLLM, config.logger, config.instanceUrl),
            (0, get_node_parameter_tool_1.createGetNodeParameterTool)(),
            (0, validate_configuration_tool_1.createValidateConfigurationTool)(config.parsedNodeTypes),
            ...(config.resourceLocatorCallback
                ? [
                    (0, get_resource_locator_options_tool_1.createGetResourceLocatorOptionsTool)(config.parsedNodeTypes, config.resourceLocatorCallback, config.logger),
                ]
                : []),
        ];
        const tools = includeExamples
            ? [...baseTools, (0, get_node_examples_tool_1.createGetNodeConfigurationExamplesTool)(config.logger)]
            : baseTools;
        this.toolMap = new Map(tools.map((bt) => [bt.tool.name, bt.tool]));
        const systemPromptTemplate = prompts_1.ChatPromptTemplate.fromMessages([
            [
                'system',
                [
                    {
                        type: 'text',
                        text: (0, configurator_prompt_1.buildConfiguratorPrompt)(),
                    },
                    {
                        type: 'text',
                        text: configurator_prompt_1.INSTANCE_URL_PROMPT,
                        cache_control: { type: 'ephemeral' },
                    },
                ],
            ],
            ['placeholder', '{messages}'],
        ]);
        if (typeof config.llm.bindTools !== 'function') {
            throw new errors_1.LLMServiceError('LLM does not support tools', {
                llmModel: config.llm._llmType(),
            });
        }
        this.agent = systemPromptTemplate.pipe(config.llm.bindTools(tools.map((bt) => bt.tool)));
        const prefetchRLCNode = async (state) => {
            if (!this.resourceLocatorCallback) {
                return { prefetchedRLCOptions: [] };
            }
            const parametersToFetch = (0, rlc_prefetch_1.detectRLCParametersForPrefetch)(state.workflowJSON.nodes, this.nodeTypes);
            if (parametersToFetch.length === 0) {
                return { prefetchedRLCOptions: [] };
            }
            this.logger?.debug('Pre-fetching RLC options', {
                parameterCount: parametersToFetch.length,
                parameters: parametersToFetch.map((p) => `${p.nodeName}.${p.parameterPath}`),
            });
            const results = await (0, rlc_prefetch_1.prefetchRLCOptions)(parametersToFetch, state.workflowJSON.nodes, this.resourceLocatorCallback, this.logger);
            this.logger?.debug('RLC prefetch completed', {
                successCount: results.filter((r) => !r.error).length,
                errorCount: results.filter((r) => r.error).length,
            });
            const formattedOptions = (0, rlc_prefetch_1.formatPrefetchedOptionsForLLM)(results);
            if (formattedOptions && state.messages.length > 0) {
                const firstMessage = state.messages[0];
                if (typeof firstMessage.content === 'string') {
                    firstMessage.content = `${firstMessage.content}\n\n${formattedOptions}`;
                }
            }
            return { prefetchedRLCOptions: results };
        };
        const callAgent = async (state) => {
            (0, cache_control_1.applySubgraphCacheMarkers)(state.messages);
            const response = await this.agent.invoke({
                messages: state.messages,
                instanceUrl: state.instanceUrl ?? '',
            });
            if (!(0, langchain_1.isBaseMessage)(response)) {
                throw new errors_1.LLMServiceError('Configurator agent did not return a valid message');
            }
            return { messages: [response] };
        };
        const subgraph = new langgraph_1.StateGraph(exports.ConfiguratorSubgraphState)
            .addNode('prefetch_rlc', prefetchRLCNode)
            .addNode('agent', callAgent)
            .addNode('tools', async (state) => await (0, subgraph_helpers_1.executeSubgraphTools)(state, this.toolMap))
            .addNode('process_operations', operations_processor_1.processOperations)
            .addEdge('__start__', 'prefetch_rlc')
            .addEdge('prefetch_rlc', 'agent')
            .addConditionalEdges('agent', (0, subgraph_helpers_1.createStandardShouldContinue)())
            .addEdge('tools', 'process_operations')
            .addEdge('process_operations', 'agent');
        return subgraph.compile();
    }
    transformInput(parentState) {
        const userRequest = (0, subgraph_helpers_1.extractUserRequest)(parentState.messages);
        const contextParts = [];
        if (userRequest) {
            contextParts.push('=== USER REQUEST ===');
            contextParts.push(userRequest);
        }
        if (parentState.discoveryContext) {
            contextParts.push('=== DISCOVERY CONTEXT ===');
            contextParts.push((0, context_builders_1.buildDiscoveryContextBlock)(parentState.discoveryContext, true));
        }
        const builderErrorEntry = parentState.coordinationLog?.find((entry) => {
            if (entry.status !== 'error')
                return false;
            if (entry.phase !== 'builder')
                return false;
            return (entry.metadata.phase === 'error' &&
                'partialBuilderData' in entry.metadata &&
                entry.metadata.partialBuilderData);
        });
        if (builderErrorEntry?.metadata.phase === 'error' &&
            builderErrorEntry.metadata.partialBuilderData) {
            const { nodeCount, nodeNames } = builderErrorEntry.metadata.partialBuilderData;
            contextParts.push((0, configurator_prompt_1.buildRecoveryModeContext)(nodeCount, nodeNames));
        }
        contextParts.push('=== WORKFLOW TO CONFIGURE ===');
        contextParts.push((0, context_builders_1.buildWorkflowJsonBlock)(parentState.workflowJSON));
        contextParts.push('=== EXECUTION CONTEXT ===');
        contextParts.push((0, context_builders_1.buildExecutionContextBlock)(parentState.workflowContext));
        const contextMessage = (0, context_builders_1.createContextMessage)(contextParts);
        return {
            workflowJSON: parentState.workflowJSON,
            workflowContext: parentState.workflowContext,
            instanceUrl: this.instanceUrl,
            userRequest,
            discoveryContext: parentState.discoveryContext,
            messages: [contextMessage],
            cachedTemplates: parentState.cachedTemplates,
        };
    }
    transformOutput(subgraphOutput, _parentState) {
        const lastMessage = subgraphOutput.messages[subgraphOutput.messages.length - 1];
        const setupInstructions = typeof lastMessage?.content === 'string' ? lastMessage.content : 'Configuration complete';
        const nodesConfigured = subgraphOutput.workflowJSON.nodes.length;
        const hasSetupInstructions = setupInstructions.includes('Setup') ||
            setupInstructions.includes('setup') ||
            setupInstructions.length > 50;
        const logEntry = {
            phase: 'configurator',
            status: 'completed',
            timestamp: Date.now(),
            summary: `Configured ${nodesConfigured} nodes`,
            output: setupInstructions,
            metadata: (0, coordination_1.createConfiguratorMetadata)({
                nodesConfigured,
                hasSetupInstructions,
            }),
        };
        return {
            workflowJSON: subgraphOutput.workflowJSON,
            workflowOperations: subgraphOutput.workflowOperations ?? [],
            coordinationLog: [logEntry],
            cachedTemplates: subgraphOutput.cachedTemplates,
        };
    }
}
exports.ConfiguratorSubgraph = ConfiguratorSubgraph;
