// @n8n/ai-workflow-builder/dist/subgraphs/discovery.subgraph.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscoverySubgraph = exports.DiscoverySubgraphState = void 0;
const messages_1 = require("@langchain/core/messages");
const prompts_1 = require("@langchain/core/prompts");
const tools_1 = require("@langchain/core/tools");
const langgraph_1 = require("@langchain/langgraph");
const zod_1 = require("zod");
const errors_1 = require("../errors");
const discovery_prompt_1 = require("../prompts/agents/discovery.prompt");
const resource_operation_extractor_1 = require("../utils/resource-operation-extractor");
const subgraph_interface_1 = require("./subgraph-interface");
const get_documentation_tool_1 = require("../tools/get-documentation.tool");
const get_workflow_examples_tool_1 = require("../tools/get-workflow-examples.tool");
const node_details_tool_1 = require("../tools/node-details.tool");
const node_search_tool_1 = require("../tools/node-search.tool");
const coordination_1 = require("../types/coordination");
const cache_control_1 = require("../utils/cache-control");
const context_builders_1 = require("../utils/context-builders");
const state_reducers_1 = require("../utils/state-reducers");
const subgraph_helpers_1 = require("../utils/subgraph-helpers");
const discoveryOutputSchema = zod_1.z.object({
    nodesFound: zod_1.z
        .array(zod_1.z.object({
        nodeName: zod_1.z.string().describe('The internal name of the node (e.g., n8n-nodes-base.gmail)'),
        version: zod_1.z
            .number()
            .describe('The version number of the node (e.g., 1, 1.1, 2, 3, 3.2, etc.)'),
        reasoning: zod_1.z.string().describe('Why this node is relevant for the workflow'),
        connectionChangingParameters: zod_1.z
            .array(zod_1.z.object({
            name: zod_1.z
                .string()
                .describe('Parameter name (e.g., "mode", "operation", "hasOutputParser")'),
            possibleValues: zod_1.z
                .array(zod_1.z.union([zod_1.z.string(), zod_1.z.boolean(), zod_1.z.number()]))
                .describe('Possible values this parameter can take'),
        }))
            .describe('Parameters that affect node connections (inputs/outputs). ONLY include if parameter appears in <input> or <output> expressions'),
    }))
        .describe('List of n8n nodes identified as necessary for the workflow'),
});
exports.DiscoverySubgraphState = langgraph_1.Annotation.Root({
    userRequest: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => '',
    }),
    messages: (0, langgraph_1.Annotation)({
        reducer: (x, y) => x.concat(y),
        default: () => [],
    }),
    nodesFound: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => [],
    }),
    bestPractices: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
    }),
    templateIds: (0, langgraph_1.Annotation)({
        reducer: state_reducers_1.appendArrayReducer,
        default: () => [],
    }),
    cachedTemplates: (0, langgraph_1.Annotation)({
        reducer: state_reducers_1.cachedTemplatesReducer,
        default: () => [],
    }),
    resourceOperationCache: (0, langgraph_1.Annotation)({
        reducer: (x, y) => ({ ...x, ...y }),
        default: () => ({}),
    }),
});
class DiscoverySubgraph extends subgraph_interface_1.BaseSubgraph {
    name = 'discovery_subgraph';
    description = 'Discovers nodes and context for the workflow';
    agent;
    toolMap;
    logger;
    parsedNodeTypes;
    create(config) {
        this.logger = config.logger;
        this.parsedNodeTypes = config.parsedNodeTypes;
        const includeExamples = config.featureFlags?.templateExamples === true;
        const baseTools = [
            (0, get_documentation_tool_1.createGetDocumentationTool)(),
            (0, node_search_tool_1.createNodeSearchTool)(config.parsedNodeTypes),
            (0, node_details_tool_1.createNodeDetailsTool)(config.parsedNodeTypes, config.logger),
        ];
        const tools = includeExamples
            ? [...baseTools, (0, get_workflow_examples_tool_1.createGetWorkflowExamplesTool)(config.logger)]
            : baseTools;
        this.toolMap = new Map(tools.map((bt) => [bt.tool.name, bt.tool]));
        const submitTool = (0, tools_1.tool)(() => { }, {
            name: 'submit_discovery_results',
            description: 'Submit the final discovery results',
            schema: discoveryOutputSchema,
        });
        const discoveryPrompt = (0, discovery_prompt_1.buildDiscoveryPrompt)({ includeExamples });
        const systemPrompt = prompts_1.ChatPromptTemplate.fromMessages([
            [
                'system',
                [
                    {
                        type: 'text',
                        text: discoveryPrompt,
                        cache_control: { type: 'ephemeral' },
                    },
                ],
            ],
            ['human', '{prompt}'],
            ['placeholder', '{messages}'],
        ]);
        if (typeof config.llm.bindTools !== 'function') {
            throw new errors_1.LLMServiceError('LLM does not support tools', {
                llmModel: config.llm._llmType(),
            });
        }
        const allTools = [...tools.map((bt) => bt.tool), submitTool];
        this.agent = systemPrompt.pipe(config.llm.bindTools(allTools));
        const subgraph = new langgraph_1.StateGraph(exports.DiscoverySubgraphState)
            .addNode('agent', this.callAgent.bind(this))
            .addNode('tools', async (state) => await (0, subgraph_helpers_1.executeSubgraphTools)(state, this.toolMap))
            .addNode('format_output', this.formatOutput.bind(this))
            .addEdge('__start__', 'agent')
            .addConditionalEdges('agent', this.shouldContinue.bind(this), {
            tools: 'tools',
            format_output: 'format_output',
            end: langgraph_1.END,
        })
            .addEdge('tools', 'agent')
            .addEdge('format_output', langgraph_1.END);
        return subgraph.compile();
    }
    async callAgent(state) {
        if (state.messages.length > 0) {
            (0, cache_control_1.applySubgraphCacheMarkers)(state.messages);
        }
        const response = (await this.agent.invoke({
            messages: state.messages,
            prompt: state.userRequest,
            techniques: (0, discovery_prompt_1.formatTechniqueList)(),
            exampleCategorizations: (0, discovery_prompt_1.formatExampleCategorizations)(),
        }));
        return { messages: [response] };
    }
    formatOutput(state) {
        const lastMessage = state.messages.at(-1);
        let output;
        if (lastMessage && (0, messages_1.isAIMessage)(lastMessage) && lastMessage.tool_calls) {
            const submitCall = lastMessage.tool_calls.find((tc) => tc.name === 'submit_discovery_results');
            if (submitCall) {
                const parseResult = discoveryOutputSchema.safeParse(submitCall.args);
                if (!parseResult.success) {
                    this.logger?.error('[Discovery] Invalid discovery output schema - returning empty results', {
                        errors: parseResult.error.errors,
                        lastMessageContent: typeof lastMessage?.content === 'string'
                            ? lastMessage.content.substring(0, 200)
                            : JSON.stringify(lastMessage?.content)?.substring(0, 200),
                    });
                    return {
                        nodesFound: [],
                        templateIds: [],
                    };
                }
                output = parseResult.data;
            }
        }
        if (!output) {
            this.logger?.error('[Discovery] No submit_discovery_results tool call found - agent may have stopped early', {
                messageCount: state.messages.length,
                lastMessageType: lastMessage?.getType(),
            });
            return {
                nodesFound: [],
                templateIds: [],
            };
        }
        const nodeTypeMap = new Map();
        for (const nt of this.parsedNodeTypes) {
            const versions = Array.isArray(nt.version) ? nt.version : [nt.version];
            for (const v of versions) {
                nodeTypeMap.set(`${nt.name}:${v}`, nt);
            }
        }
        const existingCache = state.resourceOperationCache ?? {};
        const hydratedNodesFound = output.nodesFound.map((node) => {
            const cacheKey = (0, resource_operation_extractor_1.createResourceCacheKey)(node.nodeName, node.version);
            if (cacheKey in existingCache) {
                const cached = existingCache[cacheKey];
                if (cached) {
                    return {
                        ...node,
                        availableResources: cached.resources,
                    };
                }
                return node;
            }
            const nodeType = nodeTypeMap.get(cacheKey);
            if (!nodeType) {
                this.logger?.warn('[Discovery] Node type not found during resource hydration', {
                    nodeName: node.nodeName,
                    nodeVersion: node.version,
                });
                return node;
            }
            const resourceOpInfo = (0, resource_operation_extractor_1.extractResourceOperations)(nodeType, node.version, this.logger);
            if (!resourceOpInfo) {
                return node;
            }
            return {
                ...node,
                availableResources: resourceOpInfo.resources,
            };
        });
        return {
            nodesFound: hydratedNodesFound,
            bestPractices: state.bestPractices,
            templateIds: state.templateIds ?? [],
        };
    }
    shouldContinue(state) {
        const lastMessage = state.messages[state.messages.length - 1];
        if (lastMessage &&
            (0, messages_1.isAIMessage)(lastMessage) &&
            lastMessage.tool_calls &&
            lastMessage.tool_calls.length > 0) {
            const submitCall = lastMessage.tool_calls.find((tc) => tc.name === 'submit_discovery_results');
            if (submitCall) {
                return 'format_output';
            }
            return 'tools';
        }
        this.logger?.warn('[Discovery] Agent stopped without calling submit_discovery_results - check if LLM is producing valid tool calls');
        return 'end';
    }
    transformInput(parentState) {
        const userRequest = (0, subgraph_helpers_1.extractUserRequest)(parentState.messages, 'Build a workflow');
        const contextParts = [];
        contextParts.push('<user_request>');
        contextParts.push(userRequest);
        contextParts.push('</user_request>');
        if (parentState.workflowJSON.nodes.length > 0) {
            contextParts.push('<existing_workflow_summary>');
            contextParts.push((0, context_builders_1.buildWorkflowSummary)(parentState.workflowJSON));
            contextParts.push('</existing_workflow_summary>');
        }
        const contextMessage = (0, context_builders_1.createContextMessage)(contextParts);
        return {
            userRequest,
            messages: [contextMessage],
            cachedTemplates: parentState.cachedTemplates,
        };
    }
    transformOutput(subgraphOutput, _parentState) {
        const nodesFound = subgraphOutput.nodesFound || [];
        const templateIds = subgraphOutput.templateIds || [];
        const discoveryContext = {
            nodesFound,
            bestPractices: subgraphOutput.bestPractices,
        };
        const logEntry = {
            phase: 'discovery',
            status: 'completed',
            timestamp: Date.now(),
            summary: `Discovered ${nodesFound.length} nodes`,
            metadata: (0, coordination_1.createDiscoveryMetadata)({
                nodesFound: nodesFound.length,
                nodeTypes: nodesFound.map((n) => n.nodeName),
                hasBestPractices: !!subgraphOutput.bestPractices,
            }),
        };
        return {
            discoveryContext,
            coordinationLog: [logEntry],
            templateIds,
            cachedTemplates: subgraphOutput.cachedTemplates,
        };
    }
}
exports.DiscoverySubgraph = DiscoverySubgraph;
