// @n8n/n8n-nodes-langchain/dist/nodes/mcp/McpTrigger/McpServer.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _a, _McpServerManager_instance;
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpServerManager = void 0;
const index_js_1 = require("@modelcontextprotocol/sdk/server/index.js");
const types_js_1 = require("@modelcontextprotocol/sdk/types.js");
const crypto_1 = require("crypto");
const n8n_workflow_1 = require("n8n-workflow");
const zod_to_json_schema_1 = require("zod-to-json-schema");
const FlushingTransport_1 = require("./FlushingTransport");
function wasToolCall(body) {
    try {
        const message = JSON.parse(body);
        const parsedMessage = types_js_1.JSONRPCMessageSchema.parse(message);
        return ('method' in parsedMessage &&
            'id' in parsedMessage &&
            parsedMessage?.method === types_js_1.CallToolRequestSchema.shape.method.value);
    }
    catch {
        return false;
    }
}
function getRequestId(message) {
    try {
        const parsedMessage = types_js_1.JSONRPCMessageSchema.parse(message);
        return 'id' in parsedMessage ? String(parsedMessage.id) : undefined;
    }
    catch {
        return undefined;
    }
}
class McpServerManager {
    constructor(logger) {
        this.servers = {};
        this.transports = {};
        this.tools = {};
        this.resolveFunctions = {};
        this.logger = logger;
        this.logger.debug('MCP Server created');
    }
    static instance(logger) {
        if (!__classPrivateFieldGet(_a, _a, "f", _McpServerManager_instance)) {
            __classPrivateFieldSet(_a, _a, new _a(logger), "f", _McpServerManager_instance);
            logger.debug('Created singleton MCP manager');
        }
        return __classPrivateFieldGet(_a, _a, "f", _McpServerManager_instance);
    }
    async createServerWithSSETransport(serverName, postUrl, resp) {
        const server = new index_js_1.Server({
            name: serverName,
            version: '0.1.0',
        }, {
            capabilities: {
                tools: {},
            },
        });
        const transport = new FlushingTransport_1.FlushingSSEServerTransport(postUrl, resp);
        this.setUpHandlers(server);
        const sessionId = transport.sessionId;
        this.transports[sessionId] = transport;
        this.servers[sessionId] = server;
        resp.on('close', async () => {
            this.logger.debug(`Deleting transport for ${sessionId}`);
            delete this.tools[sessionId];
            delete this.transports[sessionId];
            delete this.servers[sessionId];
        });
        await server.connect(transport);
        if (resp.flush) {
            resp.flush();
        }
    }
    getSessionId(req) {
        return (req.query.sessionId ?? req.headers['mcp-session-id']);
    }
    getTransport(sessionId) {
        return this.transports[sessionId];
    }
    async createServerWithStreamableHTTPTransport(serverName, resp, req) {
        const server = new index_js_1.Server({
            name: serverName,
            version: '0.1.0',
        }, {
            capabilities: {
                tools: {},
            },
        });
        const transport = new FlushingTransport_1.FlushingStreamableHTTPTransport({
            sessionIdGenerator: () => (0, crypto_1.randomUUID)(),
            onsessioninitialized: (sessionId) => {
                this.logger.debug(`New session initialized: ${sessionId}`);
                transport.onclose = () => {
                    this.logger.debug(`Deleting transport for ${sessionId}`);
                    delete this.tools[sessionId];
                    delete this.transports[sessionId];
                    delete this.servers[sessionId];
                };
                this.transports[sessionId] = transport;
                this.servers[sessionId] = server;
            },
        }, resp);
        this.setUpHandlers(server);
        await server.connect(transport);
        await transport.handleRequest(req, resp, req?.body);
        if (resp.flush) {
            resp.flush();
        }
    }
    async handlePostMessage(req, resp, connectedTools) {
        const sessionId = this.getSessionId(req);
        const transport = this.getTransport(sessionId);
        if (sessionId && transport) {
            const message = (0, n8n_workflow_1.jsonParse)(req.rawBody.toString());
            const messageId = getRequestId(message);
            const callId = messageId ? `${sessionId}_${messageId}` : sessionId;
            this.tools[sessionId] = connectedTools;
            try {
                await new Promise(async (resolve) => {
                    this.resolveFunctions[callId] = resolve;
                    await transport.handleRequest(req, resp, message);
                });
            }
            finally {
                delete this.resolveFunctions[callId];
            }
        }
        else {
            this.logger.warn(`No transport found for session ${sessionId}`);
            resp.status(401).send('No transport found for sessionId');
        }
        if (resp.flush) {
            resp.flush();
        }
        return wasToolCall(req.rawBody.toString());
    }
    async handleDeleteRequest(req, resp) {
        const sessionId = this.getSessionId(req);
        if (!sessionId) {
            resp.status(400).send('No sessionId provided');
            return;
        }
        const transport = this.getTransport(sessionId);
        if (transport) {
            if (transport instanceof FlushingTransport_1.FlushingStreamableHTTPTransport) {
                await transport.handleRequest(req, resp);
                return;
            }
            else {
                resp.status(405).send('Method Not Allowed');
                return;
            }
        }
        resp.status(404).send('Session not found');
    }
    setUpHandlers(server) {
        server.setRequestHandler(types_js_1.ListToolsRequestSchema, async (_, extra) => {
            if (!extra.sessionId) {
                throw new n8n_workflow_1.OperationalError('Require a sessionId for the listing of tools');
            }
            return {
                tools: this.tools[extra.sessionId].map((tool) => {
                    return {
                        name: tool.name,
                        description: tool.description,
                        inputSchema: (0, zod_to_json_schema_1.zodToJsonSchema)(tool.schema, { removeAdditionalStrategy: 'strict' }),
                    };
                }),
            };
        });
        server.setRequestHandler(types_js_1.CallToolRequestSchema, async (request, extra) => {
            if (!request.params?.name || !request.params?.arguments) {
                throw new n8n_workflow_1.OperationalError('Require a name and arguments for the tool call');
            }
            if (!extra.sessionId) {
                throw new n8n_workflow_1.OperationalError('Require a sessionId for the tool call');
            }
            const callId = extra.requestId ? `${extra.sessionId}_${extra.requestId}` : extra.sessionId;
            const requestedTool = this.tools[extra.sessionId].find((tool) => tool.name === request.params.name);
            if (!requestedTool) {
                throw new n8n_workflow_1.OperationalError('Tool not found');
            }
            try {
                const result = await requestedTool.invoke(request.params.arguments);
                if (this.resolveFunctions[callId]) {
                    this.resolveFunctions[callId]();
                }
                else {
                    this.logger.warn(`No resolve function found for ${callId}`);
                }
                this.logger.debug(`Got request for ${requestedTool.name}, and executed it.`);
                if (typeof result === 'object') {
                    return { content: [{ type: 'text', text: JSON.stringify(result) }] };
                }
                if (typeof result === 'string') {
                    return { content: [{ type: 'text', text: result }] };
                }
                return { content: [{ type: 'text', text: String(result) }] };
            }
            catch (error) {
                this.logger.error(`Error while executing Tool ${requestedTool.name}: ${error}`);
                return { isError: true, content: [{ type: 'text', text: `Error: ${error.message}` }] };
            }
        });
        server.onclose = () => {
            this.logger.debug('Closing MCP Server');
        };
        server.onerror = (error) => {
            this.logger.error(`MCP Error: ${error}`);
        };
    }
}
exports.McpServerManager = McpServerManager;
_a = McpServerManager;
_McpServerManager_instance = { value: void 0 };
