// n8n-nodes-base/dist/nodes/Currents/CurrentsTriggerHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateWebhookSecret = generateWebhookSecret;
exports.listWebhooks = listWebhooks;
exports.findWebhookByUrl = findWebhookByUrl;
exports.createWebhook = createWebhook;
exports.updateWebhook = updateWebhook;
exports.deleteWebhook = deleteWebhook;
exports.verifyWebhook = verifyWebhook;
exports.isTimestampValid = isTimestampValid;
const crypto_1 = require("crypto");
const CURRENTS_API_BASE = 'https://api.currents.dev/v1';
/**
 * Maximum allowed age for a webhook request timestamp (5 minutes).
 * Requests older than this are considered potential replay attacks.
 */
const MAX_TIMESTAMP_AGE_SECONDS = 300;
/**
 * Header name used for webhook secret validation.
 */
const WEBHOOK_SECRET_HEADER = 'x-webhook-secret';
/**
 * Generates a cryptographically secure random secret for webhook validation.
 */
function generateWebhookSecret() {
    return (0, crypto_1.randomBytes)(32).toString('hex');
}
/**
 * Lists all webhooks for a project.
 */
async function listWebhooks(projectId) {
    const response = await this.helpers.httpRequestWithAuthentication.call(this, 'currentsApi', {
        method: 'GET',
        url: `${CURRENTS_API_BASE}/webhooks`,
        qs: { projectId },
    });
    return response.data ?? [];
}
/**
 * Finds an existing webhook by URL for a project.
 */
async function findWebhookByUrl(projectId, webhookUrl) {
    const webhooks = await listWebhooks.call(this, projectId);
    return webhooks.find((webhook) => webhook.url === webhookUrl);
}
/**
 * Creates a new webhook in Currents.
 */
async function createWebhook(projectId, options) {
    const response = await this.helpers.httpRequestWithAuthentication.call(this, 'currentsApi', {
        method: 'POST',
        url: `${CURRENTS_API_BASE}/webhooks`,
        qs: { projectId },
        body: {
            url: options.url,
            hookEvents: options.hookEvents ?? [],
            headers: options.headers,
            label: options.label,
        },
    });
    return response.data;
}
/**
 * Updates an existing webhook in Currents.
 */
async function updateWebhook(hookId, options) {
    const response = await this.helpers.httpRequestWithAuthentication.call(this, 'currentsApi', {
        method: 'PUT',
        url: `${CURRENTS_API_BASE}/webhooks/${hookId}`,
        body: {
            ...(options.url && { url: options.url }),
            ...(options.hookEvents && { hookEvents: options.hookEvents }),
            ...(options.headers && { headers: options.headers }),
            ...(options.label && { label: options.label }),
        },
    });
    return response.data;
}
/**
 * Deletes a webhook from Currents.
 */
async function deleteWebhook(hookId) {
    await this.helpers.httpRequestWithAuthentication.call(this, 'currentsApi', {
        method: 'DELETE',
        url: `${CURRENTS_API_BASE}/webhooks/${hookId}`,
    });
}
/**
 * Verifies the webhook request is recent and validates the secret.
 *
 * Uses auto-managed secret from workflow static data.
 *
 * Currents.dev includes an `x-timestamp` header with the epoch timestamp in milliseconds.
 * This function validates that the timestamp is within an acceptable window to prevent
 * replay attacks.
 *
 * @returns true if the request is valid, false otherwise
 */
function verifyWebhook() {
    const req = this.getRequestObject();
    const headerData = this.getHeaderData();
    // Check timestamp to prevent replay attacks (Currents sends milliseconds)
    const timestampHeader = req.headers['x-timestamp'];
    if (typeof timestampHeader === 'string') {
        const requestTimeMs = parseInt(timestampHeader, 10);
        if (isNaN(requestTimeMs)) {
            return false;
        }
        const requestTimeSec = Math.floor(requestTimeMs / 1000);
        const currentTimeSec = Math.floor(Date.now() / 1000);
        const age = Math.abs(currentTimeSec - requestTimeSec);
        if (age > MAX_TIMESTAMP_AGE_SECONDS) {
            return false;
        }
    }
    const webhookData = this.getWorkflowStaticData('node');
    const expectedSecret = webhookData.webhookSecret;
    if (typeof expectedSecret === 'string') {
        const actualSecret = headerData[WEBHOOK_SECRET_HEADER];
        if (typeof actualSecret !== 'string') {
            return false;
        }
        // Use constant-time comparison to prevent timing attacks
        if (expectedSecret.length !== actualSecret.length ||
            !(0, crypto_1.timingSafeEqual)(Buffer.from(expectedSecret), Buffer.from(actualSecret))) {
            return false;
        }
    }
    return true;
}
/**
 * Validates that a millisecond timestamp is within the acceptable window.
 * Exported separately for unit testing.
 */
function isTimestampValid(timestampMs, currentTimeSec) {
    const requestTimeSec = Math.floor(timestampMs / 1000);
    const now = currentTimeSec ?? Math.floor(Date.now() / 1000);
    const age = Math.abs(now - requestTimeSec);
    return age <= MAX_TIMESTAMP_AGE_SECONDS;
}
