// @n8n/n8n-nodes-langchain/dist/nodes/Guardrails/v2/GuardrailsV2.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuardrailsV2 = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const execute_1 = require("../actions/execute");
const description_1 = require("../description");
const configureNodeInputs_1 = require("../helpers/configureNodeInputs");
class GuardrailsV2 {
    constructor(baseDescription) {
        this.description = {
            ...baseDescription,
            version: [2],
            inputs: `={{(${configureNodeInputs_1.configureNodeInputsV2})($parameter)}}`,
            outputs: `={{
		((parameters) => {
			const operation = parameters.operation ?? 'classify';

			if (operation === 'classify') {
				return [{displayName: "Pass", type: "${n8n_workflow_1.NodeConnectionTypes.Main}"}, {displayName: "Fail", type: "${n8n_workflow_1.NodeConnectionTypes.Main}"}]
			}

			return [{ displayName: "", type: "${n8n_workflow_1.NodeConnectionTypes.Main}"}]
		})($parameter)
	}}`,
            defaults: {
                name: 'Guardrails',
            },
            properties: description_1.propertiesDescription,
        };
    }
    async execute() {
        return await execute_1.execute.call(this);
    }
}
exports.GuardrailsV2 = GuardrailsV2;
