// @n8n/api-types/dist/dto/variables/base.dto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseVariableRequestDto = exports.variableTypeSchema = exports.variableValueSchema = exports.variableKeySchema = exports.TYPE_DEFAULT = exports.TYPE_ENUM = exports.VALUE_MAX_LENGTH = exports.KEY_MAX_LENGTH = exports.KEY_NAME_REGEX = void 0;
const zod_1 = require("zod");
const zod_class_1 = require("zod-class");
exports.KEY_NAME_REGEX = /^[A-Za-z0-9_]+$/;
exports.KEY_MAX_LENGTH = 50;
exports.VALUE_MAX_LENGTH = 1000;
exports.TYPE_ENUM = ['string'];
exports.TYPE_DEFAULT = 'string';
exports.variableKeySchema = zod_1.z
    .string()
    .min(1, 'key must be at least 1 character long')
    .max(exports.KEY_MAX_LENGTH, 'key cannot be longer than 50 characters')
    .regex(exports.KEY_NAME_REGEX, 'key can only contain characters A-Za-z0-9_');
exports.variableValueSchema = zod_1.z
    .string()
    .max(exports.VALUE_MAX_LENGTH, `value cannot be longer than ${exports.VALUE_MAX_LENGTH} characters`);
exports.variableTypeSchema = zod_1.z.enum(exports.TYPE_ENUM).default(exports.TYPE_DEFAULT);
class BaseVariableRequestDto extends zod_class_1.Z.class({
    key: exports.variableKeySchema,
    type: exports.variableTypeSchema,
    value: exports.variableValueSchema,
}) {
}
exports.BaseVariableRequestDto = BaseVariableRequestDto;
