// n8n/dist/modules/external-secrets.ee/secrets-providers-types.controller.ee.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretProvidersTypesController = void 0;
const backend_common_1 = require("@n8n/backend-common");
const decorators_1 = require("@n8n/decorators");
const bad_request_error_1 = require("../../errors/response-errors/bad-request.error");
const external_secrets_config_1 = require("./external-secrets.config");
let SecretProvidersTypesController = class SecretProvidersTypesController {
    constructor(config, logger) {
        this.config = config;
        this.logger = logger;
        this.logger = this.logger.scoped('external-secrets');
    }
    checkFeatureFlag(_req, _res, next) {
        if (!this.config.externalSecretsForProjects) {
            throw new bad_request_error_1.BadRequestError('External secrets for projects feature is not enabled');
        }
        next();
    }
    async listSecretProviderTypes() {
        this.logger.debug('List provider connection types');
        return await Promise.resolve([]);
    }
    async getSecretProviderType() {
        this.logger.debug('Get provider connection type');
        return await Promise.resolve({});
    }
};
exports.SecretProvidersTypesController = SecretProvidersTypesController;
__decorate([
    (0, decorators_1.Middleware)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Function]),
    __metadata("design:returntype", void 0)
], SecretProvidersTypesController.prototype, "checkFeatureFlag", null);
__decorate([
    (0, decorators_1.Get)('/'),
    (0, decorators_1.GlobalScope)('externalSecretsProvider:list'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], SecretProvidersTypesController.prototype, "listSecretProviderTypes", null);
__decorate([
    (0, decorators_1.Get)('/:type'),
    (0, decorators_1.GlobalScope)('externalSecretsProvider:list'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], SecretProvidersTypesController.prototype, "getSecretProviderType", null);
exports.SecretProvidersTypesController = SecretProvidersTypesController = __decorate([
    (0, decorators_1.RestController)('/secret-providers/types'),
    __metadata("design:paramtypes", [external_secrets_config_1.ExternalSecretsConfig,
        backend_common_1.Logger])
], SecretProvidersTypesController);
