// n8n/dist/modules/external-secrets.ee/secrets-providers-connections.service.ee.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsProvidersConnectionsService = void 0;
const db_1 = require("@n8n/db");
const di_1 = require("@n8n/di");
const n8n_core_1 = require("n8n-core");
const bad_request_error_1 = require("../../errors/response-errors/bad-request.error");
const not_found_error_1 = require("../../errors/response-errors/not-found.error");
let SecretsProvidersConnectionsService = class SecretsProvidersConnectionsService {
    constructor(repository, projectAccessRepository, cipher) {
        this.repository = repository;
        this.projectAccessRepository = projectAccessRepository;
        this.cipher = cipher;
    }
    async createConnection(proposedConnection) {
        const existing = await this.repository.findOne({
            where: { providerKey: proposedConnection.providerKey },
        });
        if (existing) {
            throw new bad_request_error_1.BadRequestError(`Connection with key "${proposedConnection.providerKey}" already exists`);
        }
        const encryptedSettings = this.encryptConnectionSettings(proposedConnection.settings);
        const connection = this.repository.create({
            ...proposedConnection,
            encryptedSettings,
            isEnabled: false,
        });
        const savedConnection = await this.repository.save(connection);
        if (proposedConnection.projectIds.length > 0) {
            const entries = proposedConnection.projectIds.map((projectId) => this.projectAccessRepository.create({
                secretsProviderConnectionId: savedConnection.id,
                projectId,
            }));
            await this.projectAccessRepository.save(entries);
        }
        return (await this.repository.findOne({
            where: { providerKey: proposedConnection.providerKey },
        }));
    }
    async updateConnection(providerKey, updates) {
        const connection = await this.repository.findOne({ where: { providerKey } });
        if (!connection) {
            throw new not_found_error_1.NotFoundError(`Connection with key "${providerKey}" not found`);
        }
        if (updates.type !== undefined) {
            connection.type = updates.type;
            if (!updates.settings) {
                throw new bad_request_error_1.BadRequestError('When changing the connection type, new settings must be provided as well');
            }
        }
        if (updates.settings !== undefined) {
            connection.encryptedSettings = this.encryptConnectionSettings(updates.settings);
        }
        await this.repository.save(connection);
        if (updates.projectIds !== undefined) {
            await this.projectAccessRepository.setProjectAccess(connection.id, updates.projectIds);
        }
        return (await this.repository.findOne({ where: { providerKey } }));
    }
    async deleteConnection(providerKey) {
        const connection = await this.repository.findOne({ where: { providerKey } });
        if (!connection) {
            throw new not_found_error_1.NotFoundError(`Connection with key "${providerKey}" not found`);
        }
        await this.projectAccessRepository.deleteByConnectionId(connection.id);
        await this.repository.remove(connection);
        return connection;
    }
    async getConnection(providerKey) {
        const connection = await this.repository.findOne({ where: { providerKey } });
        if (!connection) {
            throw new not_found_error_1.NotFoundError(`Connection with key "${providerKey}" not found`);
        }
        return connection;
    }
    async listConnections() {
        return await this.repository.findAll();
    }
    toPublicConnection(connection) {
        return {
            id: String(connection.id),
            name: connection.providerKey,
            type: connection.type,
            isEnabled: connection.isEnabled,
            projects: connection.projectAccess.map((access) => ({
                id: access.project.id,
                name: access.project.name,
            })),
            createdAt: connection.createdAt.toISOString(),
            updatedAt: connection.updatedAt.toISOString(),
        };
    }
    encryptConnectionSettings(settings) {
        return this.cipher.encrypt(settings);
    }
};
exports.SecretsProvidersConnectionsService = SecretsProvidersConnectionsService;
exports.SecretsProvidersConnectionsService = SecretsProvidersConnectionsService = __decorate([
    (0, di_1.Service)(),
    __metadata("design:paramtypes", [db_1.SecretsProviderConnectionRepository,
        db_1.ProjectSecretsProviderAccessRepository,
        n8n_core_1.Cipher])
], SecretsProvidersConnectionsService);
